src/common/ast-path.js

```diff
@@ -89,24 +115,55 @@ class AstPath {
       stack.push(name, value);
     }
 
-    for (let i = 0; i < value.length; ++i) {
-      stack.push(i, value[i]);
-      callback(this, i, value);
-      stack.length -= 2;
-    }
-
-    stack.length = length;
+    const iteratee = (index) =>
+      tryFinally(
+        () => {
+          stack.push(index, value[index]);
+          return callback(this, index, value);
+        },
+        () => {
+          stack.length -= 2;
+        }
+      );
+
+    return tryFinally(
+      () => {
+        let promise;
+
+        for (let index = 0; index < value.length; index++) {
+          if (promise) {
+            promise = promise.then(() => iteratee(index));
+            continue;
+          }
+
+          const result = iteratee(index);
+
+          if (isThenable(result)) {
+            promise = result;
+          }
+        }
+
+        return promise;
+      },
+      () => {
+        stack.length = length;
+      }
+    );
   }
 
   // Similar to AstPath.prototype.each, except that the results of the
   // callback function invocations are stored in an array and returned at
   // the end of the iteration.
   map(callback, ...names) {
-    const result = [];
-    this.each((path, index, value) => {
-      result[index] = callback(path, index, value);
+    const array = [];
+    const maybePromise = this.each((path, index, value) => {
```

Comment:
What is returned by `each`...?

Reply:
`Promise | undefined`

Reply:
Ideally it should be `Promise<undefined> | undefined`, but I don't think anyone expect value for this call, it returns `Promise<value  from last callback>` now, because I don't want add another `.then()` call to the final promise.

Reply:
I got it. When does `each` return `Promise` and `undefined`, respectively?

Reply:
When we see any `Thenable` that callback returns. `.each` will chain rest `callback`s and return Promise.

```js
path.each((path, index) => index === 3 ? Promise.resolve(1) : 2, "foo")
```

if this `node.foo` has less than 4 elements, the result will be `undefined`, otherwise it will be Promise.

Reply:
I did  implement `each/eachSync` at first, then I changed to current implementation, I was hoping to bring less breaking changes.